lib/index.js

```diff
@@ -12,88 +12,101 @@ const Manager = require('./manager');
+        const rootRealm = internals.getRootRealm(server.realm);
+        const rootState = internals.state(rootRealm);
 
-            const manager = new Manager(options);
-            this.realm.plugins.vision.manager = manager;
-            return manager;
-        });
+        if (rootState.setup) {
+            return;
+        }
 
+        server.decorate('server', 'views', internals.assignManager);
         server.decorate('server', 'render', internals.render);
         server.decorate('request', 'render', internals.render);
         server.decorate('handler', 'view', internals.handler);
+        server.decorate('toolkit', 'view', internals.toolkitView);
+
+        rootState.setup = true;
+    }
+};
+
 
-        server.decorate('toolkit', 'view', function (template, context, options) {
+internals.assignManager = function (options, realm) {
 
-            const realm = (this.realm.plugins.vision || this.request.server.realm.plugins.vision || {});
-            Hoek.assert(realm.manager, 'Cannot render view without a views manager configured');
-            return this.response(realm.manager._response(template, context, options, this.request));
-        });
+    realm = realm || this.realm;
+    const realmState = internals.state(realm);
+
+    Hoek.assert(!realmState.manager, 'Cannot set views manager more than once per realm');
+
+    if (!options.relativeTo &&
+        realm.settings.files.relativeTo) {
+
+        options = Hoek.shallow(options);
+        options.relativeTo = realm.settings.files.relativeTo;
     }
+
+    const manager = new Manager(options);
+    realmState.manager = manager;
+    return manager;
 };
 
 
 internals.render = async function (template, context, options = {}) {
 
     const isServer = (typeof this.route === 'function');
-    const server = (isServer ? this : this.server);
-    const vision = ((!isServer ? this.route.realm.plugins.vision : null) || internals.realm(server));
-    Hoek.assert(vision.manager, 'Missing views manager');
-    return await vision.manager.render(template, context, options);
-};
+    const realm = (isServer ? this.realm : this.route.realm);
+
+    const manager = internals.nearestManager(realm);
 
+    Hoek.assert(manager, 'Missing views manager');
+    return await manager.render(template, context, options);
+};
 
-internals.realm = function (server) {
+internals.toolkitView = function (template, context, options) {
 
-    if (server.realm.plugins.vision) {
-        return server.realm.plugins.vision;
-    }
+    let realm;
 
-    let parent = server.realm.parent;
-    while (parent) {
-        if (parent.plugins.vision) {
-            return parent.plugins.vision;
-        }
+    // This fingerprint is set when using the 'onRequest' server ext
+    if (this.request.route.fingerprint === '/#') {
 
-        parent = parent.parent;
+        // 'this' is the server that registered the extension
+        realm = this.realm;
+    }
+    else {
+        realm = this.request.route.realm;
```

Comment:
I'm curious what others think of using the fingerprint `/#` for detecting when using the 'onRequest' server extension.